File: Demos/SSAlertOCDemo/SSAlertOCDemo/ViewController.m
//
//  ViewController.m
//  SSAlertOCDemo
//
//  Created by yangsq on 2021/8/12.
//

#import "ViewController.h"
#import <SSAlert/SSAlert.h>
#import <Masonry/Masonry.h>



@interface CustomView : UIView

@end

@implementation CustomView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = UIColor.yellowColor;
        self.frame = CGRectMake(0, 0, 200, 200);
        UILabel *label1 = [[UILabel alloc]init];
        label1.text = @"frame布局挣开";
        [self addSubview:label1];
        [label1 mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(@50);
            make.centerX.equalTo(self);
        }];
    }
    return self;
}

@end


@interface AutoLayoutCustomView : UIView

@end

@implementation AutoLayoutCustomView
- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = UIColor.yellowColor;
        UILabel *label1 = [[UILabel alloc]init];
        label1.text = @"自动布局挣开";
        label1.textAlignment = NSTextAlignmentCenter;
        [self addSubview:label1];
        [label1 mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.mas_equalTo(UIEdgeInsetsZero);
            make.width.height.equalTo(@200).priorityHigh();
        }];

    }
    return self;
}
@end

@interface ViewController ()
@property(nonatomic, strong) NSArray<NSString *> *titles;
@property(nonatomic, strong) NSArray<NSString *> *actions;
@property(nonatomic, assign) BOOL isModalPresent;
@property(nonatomic, assign) BOOL isAutoLayoutCustom;
@property(nonatomic, strong) SSAlertView *alertView;
@property(nonatomic, strong) UIView *customView;
@property(nonatomic, strong) UILabel *sizeLabel;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = @"SSAlertOCDemo";

    self.titles = @[@"From Top",
                    @"From Bottom",
                    @"From Left",
                    @"From Right",
                    @"From Center",
                    @"自定义alert，类似系统UIAlertView",
                    @"自定义ActionSheet，类似系统ActionSheet",
                    @"使用自带的SSAlertCommonView自定义"];
    self.actions = @[NSStringFromSelector(@selector(fromTop)),
                     NSStringFromSelector(@selector(fromBottom)),
                     NSStringFromSelector(@selector(fromLeft)),
                     NSStringFromSelector(@selector(fromRight)),
                     NSStringFromSelector(@selector(fromCenter)),
                     NSStringFromSelector(@selector(commonAlertView)),
                     NSStringFromSelector(@selector(commonActionSheet)),
                     NSStringFromSelector(@selector(customAlertCommonView))];
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"UITableViewCell"];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]initWithTitle: @"Settings" style:UIBarButtonItemStylePlain target:self action:@selector(showSettingView)];
    // Do any additional setup after loading the view.
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"UITableViewCell"];
    cell.textLabel.text = self.titles[indexPath.row];
    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.titles.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
    [self performSelector:NSSelectorFromString(self.actions[indexPath.row])];
#pragma clang diagnostic pop
}


- (void)fromTop {
    [self showAlertView:SSAlertDefaultAnimationTopState];
}

- (void)fromBottom {
    [self showAlertView:SSAlertDefaultAnimationBottomState];
}

- (void)fromLeft {
    [self showAlertView:SSAlertDefaultAnimationLeftState];
}

- (void)fromRight {
    [self showAlertView:SSAlertDefaultAnimationRightState];
}

- (void)fromCenter {
    [self showAlertView:SSAlertDefaultAnimationCenterState];
}


- (void)commonAlertView {
    if (self.isModalPresent) {
        SSAlertView *alertView = [SSAlertView modalAlertViewWithType: SSAlertActionAlertType title:@"自定义Alert弹窗" message:@"自带自定义Alert弹窗，类似系统的UIAlertView" cancelButton:@"取消" otherButtons:@[@"确定"] fromController:self navigationViewControllerClass:UINavigationController.class clickAction:^(NSInteger index) {
            NSLog(@"%ld", (long)index);
        }];
        
        [alertView show: true];
    } else {
        SSAlertView *alertView = [SSAlertView alertViewWithType: SSAlertActionAlertType title:@"自定义Alert弹窗" message:@"自带自定义Alert弹窗，类似系统的UIAlertView" cancelButton:@"取消" otherButtons:@[@"确定"] onView:self.navigationController.view clickAction:^(NSInteger index) {
            NSLog(@"%ld", (long)index);
        }];
        [alertView show: true];
    }
    
}

- (void)customAlertCommonView {
    SSAlertAction *action1 = [[SSAlertAction alloc]initWithType:SSAlertActionSheetType];
    action1.height = 80;
    action1.title = @"自定义Action1";
    action1.titleColor = UIColor.redColor;
    action1.titleFont = [UIFont systemFontOfSize:18];
    action1.backgroundColor = UIColor.yellowColor;
    [action1 setAddAction:^{
        NSLog(@"点击了Action1");
    }];
    
    SSAlertAction *action2 = [[SSAlertAction alloc]initWithType:SSAlertActionSheetType];
    action2.height = 40;
    action2.title = @"自定义Action2";
    action2.titleColor = UIColor.yellowColor;
    action2.titleFont = [UIFont systemFontOfSize:22];
    action2.backgroundColor = UIColor.redColor;
    [action2 setAddAction:^{
        NSLog(@"点击了Action2");
    }];
    SSAlertCommonView *commonView = [[SSAlertCommonView alloc]initWithTitle: @"标题自定义SSAlertCommonView" message: @"文本文本文本文本文本文本" viewType: SSAlertActionSheetType actions:@[action1, action2]];
    commonView.backgroundColor = UIColor.whiteColor;
    SSAlertView *alertView = [[SSAlertView alloc]initWithCustomView:commonView onView:self.navigationController.view masktype:SSAlertBackgroundMaskTypeBlack];
    SSAlertDefaultAnimation *animation = [[SSAlertDefaultAnimation alloc]initWithState: SSAlertDefaultAnimationBottomState];
    alertView.animation = animation;
    [alertView show: true];
}

- (void)commonActionSheet {
    if (self.isModalPresent) {
        SSAlertView *alertView = [SSAlertView modalAlertViewWithType:SSAlertActionSheetType title:@"自定义ActionSheet弹窗" message:@"自带自定义ActionSheet弹窗，类似系统的ActionSheet" cancelButton:@"取消" otherButtons:@[@"action1", @"action2", @"action3", @"action4"] fromController:self navigationViewControllerClass:UINavigationController.class clickAction:^(NSInteger index) {
            NSLog(@"%ld", (long)index);
        }];
        [alertView show: true];
    } else {
        SSAlertView *alertView = [SSAlertView alertViewWithType:SSAlertActionSheetType title:@"自定义ActionSheet弹窗" message:@"自带自定义ActionSheet弹窗，类似系统的ActionSheet" cancelButton:@"取消" otherButtons:@[@"action1", @"action2", @"action3", @"action4"] onView:self.navigationController.view clickAction:^(NSInteger index) {
            NSLog(@"%ld", (long)index);
        }];
        [alertView show: true];
    }
    
}

- (void)modalPresentSwitch:(UISwitch *)s {
    self.isModalPresent = s.on;
}

- (void)autolayoutSwitch:(UISwitch *)s {
    self.isAutoLayoutCustom = s.on;
}

- (void)showSettingView {
    UIView *customView = [[UIView alloc]initWithFrame: CGRectMake(0, 0, CGRectGetWidth(self.view.frame), 300)];
    customView.backgroundColor = UIColor.yellowColor;
    UILabel *label1 = [[UILabel alloc]init];
    label1.text = @"模态视图弹窗";
    [customView addSubview:label1];
    [label1 mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.offset(20);
        make.top.offset(30);
    }];
    
    UISwitch *swith1 = [[UISwitch alloc]init];
    [swith1 setOn:self.isModalPresent];
    [customView addSubview:swith1];
    [swith1 addTarget:self action:@selector(modalPresentSwitch:) forControlEvents:UIControlEventTouchUpInside];
    [swith1 mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(label1.mas_right).offset(20);
        make.centerY.equalTo(label1);
    }];
    
    
    UILabel *label2 = [[UILabel alloc]init];
    label2.text = @"自动布局custom";
    [customView addSubview:label2];
    [label2 mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.offset(20);
        make.top.equalTo(label1.mas_bottom).offset(15);
    }];
    
    UISwitch *swith2 = [[UISwitch alloc]init];
    [swith2 setOn:self.isAutoLayoutCustom];
    [swith2 addTarget:self action:@selector(autolayoutSwitch:) forControlEvents:UIControlEventTouchUpInside];
    [customView addSubview:swith2];
    [swith2 mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(label2.mas_right).offset(20);
        make.centerY.equalTo(label2);
    }];
    
    SSAlertView * alertView = [[SSAlertView alloc]initWithCustomView:customView onView:self.view masktype:SSAlertBackgroundMaskTypeBlack];
    SSAlertDefaultAnimation *animation = [[SSAlertDefaultAnimation alloc]initWithState:SSAlertDefaultAnimationBottomState];
    alertView.animation = animation;
    [alertView show: true];
}


- (void)showAlertView:(SSAlertDefaultAnimationState)type {
    
    UIView * customView = [CustomView new];
    if (self.isAutoLayoutCustom) {
        customView = [[AutoLayoutCustomView alloc]init];
    }
    
    if (self.isModalPresent) {
        UIButton *pushButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [pushButton setTitle: @"push" forState:UIControlStateNormal];
        [pushButton addTarget:self action:@selector(pushAction) forControlEvents:UIControlEventTouchUpInside];
        [customView addSubview:pushButton];
        [pushButton mas_makeConstraints:^(MASConstraintMaker *make) {
            make.center.equalTo(customView);
        }];
        _alertView = [[SSAlertView alloc]initWithCustomView:customView
                                               fromViewController:self
                                        navigationControllerClass:UINavigationController.class
                                                         masktype: SSAlertBackgroundMaskTypeBlack
                                                     canPanDimiss:true];
        
    } else {
        _alertView = [[SSAlertView alloc]initWithCustomView:customView onView:self.navigationController.view masktype:SSAlertBackgroundMaskTypeBlack];
    }
    
    _sizeLabel = [UILabel new];
    _sizeLabel.text = @"200";
    [customView addSubview:_sizeLabel];
    [_sizeLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.offset(0);
    }];
    
    UISlider *slider = [[UISlider alloc]init];
    slider.minimumValue = 200;
    slider.maximumValue = self.view.frame.size.width;
    [customView addSubview:slider];
    [slider addTarget:self action:@selector(sliderAction:) forControlEvents:UIControlEventValueChanged];
    [slider mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(_sizeLabel.mas_bottom).offset(15);
        make.bottom.offset(-40);
        make.left.offset(20);
        make.right.offset(-20).priorityLow();
        make.height.offset(20).priorityLow();
    }];
    
    self.customView = customView;
    SSAlertDefaultAnimation *animation = [[SSAlertDefaultAnimation alloc]initWithState:type];
    _alertView.animation = animation;
    [_alertView show: true];
    
}

- (void)sliderAction: (UISlider *)slider {
    _sizeLabel.text = [NSString stringWithFormat:@"%f", slider.value];
    self.customView.ss_w = slider.value;
    self.customView.ss_h = slider.value;
    [self.alertView refreshFrame: false];
}

- (void)pushAction {
    [_alertView.navigationController pushViewController:[ViewController new] animated:true];
}

@end

File: Demos/SSAlertSwiftDemo/SSAlertSwiftDemo/ViewController.swift
//
//  ViewController.swift
//  SSAlertSwiftDemo
//
//  Created by yangsq on 2021/8/13.
//

import UIKit
import SSAlert
import SnapKit

class CustomView: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .yellow
        ss_w = 200
        ss_h = 200
        
        let label = UILabel()
        label.text = "frame布局挣开";
        addSubview(label)
        label.snp.makeConstraints { make in
            make.top.equalTo(50)
            make.centerX.equalToSuperview()
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class AutoLayoutCustomView: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .yellow
        
        
        let label = UILabel()
        label.text = "自动布局挣开";
        label.textAlignment = .center
        addSubview(label)
        label.snp.makeConstraints { make in
            make.edges.equalTo(0)
            make.width.height.equalTo(200).priority(.high)
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}


class ViewController: UITableViewController {
    
    let titles = ["From Top",
                  "From Bottom",
                  "From Left",
                  "From Right",
                  "From Center",
                  "自定义alert，类似系统UIAlertView",
                  "自定义ActionSheet，类似系统ActionSheet",
                  "使用自带的SSAlertCommonView自定义"]
    
    let actions = [NSStringFromSelector(#selector(fromTop)),
                   NSStringFromSelector(#selector(fromBottom)),
                   NSStringFromSelector(#selector(fromLeft)),
                   NSStringFromSelector(#selector(fromRight)),
                   NSStringFromSelector(#selector(fromCenter)),
                   NSStringFromSelector(#selector(commonAlert)),
                   NSStringFromSelector(#selector(commonActionSheet)),
                   NSStringFromSelector(#selector(customAlertCommonView))]
    
    
    lazy var sizeLabel: UILabel = {
        let label = UILabel()
        label.text = "200"
        return label
    }()
    
    lazy var slilder: UISlider = {
        let slilder = UISlider()
        slilder.minimumValue = 200
        slilder.maximumValue = Float(self.view.frame.width)
        slilder.addTarget(self, action: #selector(sliderAction(slider:)), for: .valueChanged)
        return slilder
    }()
    
    var isAutoLayoutCustom: Bool = false
    var isModalPresent: Bool = false
    var alertView: SSAlertView?
    var customView: UIView?

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationItem.title = "SSAlertSwiftDemo"
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "UITableViewCell")
        navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Settings", style: .plain, target: self, action: #selector(settingsAction))
    }

    
    @objc func fromTop() {
        showAlertView(state: .fromTop)
    }
    
    @objc func fromBottom() {
        showAlertView(state: .fromBottom)
    }
    
    @objc func fromLeft() {
        showAlertView(state: .fromLeft)
    }
    
    @objc func fromRight() {
        showAlertView(state: .fromRight)
    }
    
    @objc func fromCenter() {
        showAlertView(state: .fromCenter)
    }
    
    @objc func commonAlert() {
        if isModalPresent {
            let alertView = SSAlertView.modalAlertView(style: .alert, title: "自定义Alert弹窗", message: "自带自定义Alert弹窗，类似系统的UIAlertView",  cancelButton: "Cancel", otherButtons: ["OK"], fromViewController: self){ index in
                print(index)
            }
            alertView.show()
        } else {
            let alertView = SSAlertView.alertView(style: .alert, title: "自定义Alert弹窗", message: "自带自定义Alert弹窗，类似系统的UIAlertView",  cancelButton: "Cancel", otherButtons: ["OK"], onView: navigationController!.view){ index in
                print(index)
            }
            alertView.show()
        }
    }
    
    @objc func commonActionSheet() {
        if isModalPresent {
            let alertView = SSAlertView.modalAlertView(style: .actionSheet, title: "自定义ActionSheet弹窗", message: "自带自定义ActionSheet弹窗，类似系统的ActionSheet",  cancelButton: "Cancel", otherButtons: ["action1", "action2", "action3", "action4"], fromViewController: self) { index in
                print(index)
            }
            alertView.show()
        } else {
            let alertView = SSAlertView.alertView(style: .actionSheet, title: "自定义ActionSheet弹窗", message: "自带自定义ActionSheet弹窗，类似系统的ActionSheet",  cancelButton: "Cancel", otherButtons: ["action1", "action2", "action3", "action4"], onView: navigationController!.view) { index in
                print(index)
            }
            
            alertView.show()
        }
    }
    
    @objc func customAlertCommonView() {
        let action1 = SSAlertAction(style: .actionSheet, title: "自定义Action1") {
            print("自定义Action1")
        }
        action1.titleColor = .red
        action1.backgroundColor = .yellow
        action1.titleFont = UIFont.systemFont(ofSize: 18)
        action1.height = 80
        
        let action2 = SSAlertAction(style: .actionSheet, title: "自定义Action2") {
            print("自定义Action2")
        }
        action2.titleColor = .yellow
        action2.backgroundColor = .red
        action2.titleFont = UIFont.systemFont(ofSize: 22)
        action2.height = 55
        
        let commonView = SSAlertCommonView(title: "自定义SSAlertCommonView", message: "文本文本文本文本文本文本", style: .actionSheet, actions: [action1, action2])
        commonView.backgroundColor = .white
        let alertView = SSAlertView(customView: commonView, onView: navigationController!.view, animation: SSAlertDefaultAnmation(state: .fromBottom))
        alertView.show()
    }

    @objc func settingsAction() {
        let customView = UIView(frame: .zero)
        customView.ss_w = view.frame.width
        customView.ss_h = 300
        customView.backgroundColor = .yellow
        let label1 = UILabel()
        label1.text = "模态视图弹窗"
        customView.addSubview(label1)
        label1.snp.makeConstraints { make in
            make.left.equalTo(20)
            make.top.equalTo(30)
        }
        
        let switch1 = UISwitch()
        switch1.setOn(isModalPresent, animated: false)
        switch1.addTarget(self, action: #selector(switch1Action(s:)), for: .touchUpInside)
        customView.addSubview(switch1)
        switch1.snp.makeConstraints { make in
            make.left.equalTo(label1.snp.right).offset(20)
            make.centerY.equalTo(label1)
        }
        
        let label2 = UILabel()
        label2.text = "自动布局custom"
        customView.addSubview(label2)
        label2.snp.makeConstraints { make in
            make.left.equalTo(20)
            make.top.equalTo(label1.snp.bottom).offset(15)
        }
        
        let switch2 = UISwitch()
        switch2.setOn(isAutoLayoutCustom, animated: false)
        switch2.addTarget(self, action: #selector(switch2Action(s:)), for: .touchUpInside)
        customView.addSubview(switch2)
        switch2.snp.makeConstraints { make in
            make.left.equalTo(label2.snp.right).offset(20)
            make.centerY.equalTo(label2)
        }
        
        let alertView = SSAlertView(customView: customView, onView: navigationController!.view)
        alertView.animation = SSAlertDefaultAnmation(state: .fromBottom)
        alertView.show()
    }
    
    func showAlertView(state: SSAlertDefaultAnmation.State){
        var customView = CustomView() as UIView
        if isAutoLayoutCustom {
            customView = AutoLayoutCustomView() as UIView
        }
        if isModalPresent {
            let pushButton = UIButton(type: .system)
            pushButton.setTitle("push", for: .normal)
            pushButton.addTarget(self, action: #selector(pushAction), for: .touchUpInside)
            customView.addSubview(pushButton)
            pushButton.snp.makeConstraints { make in
                make.center.equalToSuperview()
            }
            let alertView = SSAlertView(customView: customView, fromViewController: self)
            self.alertView = alertView
        } else {
            let alertView = SSAlertView(customView: customView, onView: navigationController!.view)
            self.alertView = alertView
        }
        
        customView.addSubview(sizeLabel)
        sizeLabel.snp.makeConstraints { make in
            make.centerX.equalToSuperview()
        }
        
        customView.addSubview(slilder)
        slilder.snp.makeConstraints { make in
            make.top.equalTo(sizeLabel.snp.bottom).offset(15)
            make.bottom.equalTo(-30)
            make.left.equalTo(20)
            make.right.equalTo(-20).priority(.low)
            make.height.equalTo(20).priority(.low)
        }
        self.customView = customView
        let animation = SSAlertDefaultAnmation(state: state)
        alertView?.animation = animation
        alertView?.show()
       
        
    }
    
    @objc func pushAction() {
        alertView?.navigationController?.pushViewController(ViewController(), animated: true)
    }
    
    @objc func sliderAction(slider: UISlider) {
        sizeLabel.text = "\(slilder.value)"
        customView?.ss_w = CGFloat(slilder.value)
        customView?.ss_h = CGFloat(slilder.value)
        self.alertView?.refreshFrame(animated: false)
    }
    
    @objc func switch1Action(s: UISwitch) {
        isModalPresent = s.isOn
    }
    
    @objc func switch2Action(s: UISwitch) {
        isAutoLayoutCustom = s.isOn
    }
}

extension ViewController {
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "UITableViewCell")!
        cell.textLabel?.text = titles[indexPath.row]
        return cell
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return titles.count
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.perform(NSSelectorFromString(actions[indexPath.row]))
    }
}

